### PR TITLE
Make Video.cpp into a class, with a GetVideo() singleton in AppleWin.cpp

### PR DESCRIPTION
**Pull request by tomcw**
Currently the singleton returned by GetVideo() is just a class Video instance.
I still plan to derive WinVideo from Video, and return the WinVideo subclass instead.

### CODE REVIEW

**Comment by tomcw** [comment]
@audetto: I've create this PR early to give you visibility on what I'm doing. I could even merge it at this stage, but I want to next convert WinVideo into a class as well, and derive it from the Video class.

**Comment by tomcw** [comment]
Key changes:
- class WinVideo derives from class Video
- class Video has a handful of pure virtual functions which are implemented in WinVideo
- There are also Initialize() and Destroy() implementations in both Video & WinVideo
  - They aren't in to ctor/dtor since the singleton's dtor is only called at shutdown, whereas Destroy() is called when the Apple II virtual machine is restarted (eg. on a hardware change).

**Comment by tomcw** [comment]
NB. This supersedes PR #812.

**Review by audetto — COMMENTED** [review]
I had a couple of cosmetic changes. The rest looks good, in the right direction.

File: source/Video.h [hunk 1]
```diff
@@ -177,74 +177,170 @@ struct WinBmpHeader4_t
...
+	}
 
-VideoType_e GetVideoType(void);
-void SetVideoType(VideoType_e newVideoType);
-VideoStyle_e GetVideoStyle(void);
-void SetVideoStyle(VideoStyle_e newVideoStyle);
-bool IsVideoStyle(VideoStyle_e mask);
+	virtual void Initialize(void)
+	{
+	};
```
[the remark this answers is not in the document]
Comment (tomcw) [reply source/Video.h]: Fixed

File: source/Video.h [hunk 2]
```diff
@@ -177,74 +177,170 @@ struct WinBmpHeader4_t
...
+
+protected:
+	uint8_t *g_pFramebufferbits;
+
+private:
+	void Util_MakeScreenShotFileName(TCHAR *pFinalFileName_, DWORD chars);
+	bool Util_TestScreenShotFileName(const TCHAR *pFileName);
+	void Video_MakeScreenShot(FILE *pFile, const VideoScreenShot_e ScreenShotType);
+
+	std::string Video::VideoGetSnapshotStructName(void);
```
[the remark this answers is not in the document]
Comment (tomcw) [reply source/Video.h]: Fixed

File: source/Windows/WinVideo.h [hunk 3]
```diff
@@ -1,18 +1,48 @@
...
+	virtual void VideoRefreshScreen(uint32_t uRedrawWholeScreenVideoMode = 0, bool bRedrawWholeScreen = false);
+	virtual void Video_RedrawAndTakeScreenShot(const char* pScreenshotFilename);
+	virtual void ChooseMonochromeColor(void);
+	virtual void Benchmark(void);
+	virtual void DisplayLogo(void);
+
+private:
+	void videoCreateDIBSection(void);
+	void VideoDrawLogoBitmap(HDC hDstDC, int xoff, int yoff, int srcw, int srch, int scale);
+	static BOOL CALLBACK WinVideo::DDEnumProc(LPGUID lpGUID, LPCTSTR lpszDesc, LPCTSTR lpszDrvName, LPVOID lpContext);
```
[the remark this answers is not in the document]
Comment (tomcw) [reply source/Windows/WinVideo.h]: Fixed

File: source/Windows/WinVideo.cpp [hunk 4]
```diff
@@ -185,7 +184,7 @@ void VideoBenchmark () {
       memset(mem+0x2000,0x14,0x2000);
     else
       memcpy(mem+0x2000,mem+((cycle & 2) ? 0x4000 : 0x6000),0x2000);
-    VideoRefreshScreen();
+    GetVideo().VideoRefreshScreen();
```
[the remark this answers is not in the document]
Comment (tomcw) [reply source/Windows/WinVideo.cpp]: Fixed

**Pull request merged**